progenitor-impl/src/method.rs

```diff
@@ -2224,16 +2331,223 @@ impl Generator {
+
+        params
+            .iter()
+            .filter_map(|param| match (&param.kind, &param.typ) {
+                (
+                    OperationParameterKind::Body(BodyContentType::FormData(
+                        required,
+                    )),
+                    OperationParameterType::FormPart,
+                ) => Some((*required, &param.api_name)),
+                (
+                    OperationParameterKind::Body(BodyContentType::FormData(
+                        true,
+                    )),
+                    OperationParameterType::Type(type_id),
+                ) => {
+                    build_form_type_id = Some(type_id);
+                    None
+                }
+                _ => None,
+            })
+            .for_each(|(required, api_name)| {
+                form_data_names.insert(api_name.clone());
+                let param_ident = format_ident!("{}", api_name);
+                if required {
+                    form_parts.push(quote! { .part(#api_name, #param_ident) });
+                } else {
+                    form_parts_optional.push(quote! {
+                        if let Some(#param_ident) = #param_ident {
+                            #form_ident = #form_ident.part(#api_name, #param_ident);
+                        };
+                    });
+                };
+            });
+
+        // set params from form
+        if let Some(type_id) = build_form_type_id {
+            let typ = self.get_type_space().get_type(type_id)?;
+            let td = typ.details();
+            let typify::TypeDetails::Struct(tstru) = td else {
+                unimplemented!("form-data bodies must be a Struct")
+            };
+            let body_parts = tstru
+                .properties()
+                // skip Parts made from binary strings
+                .filter_map(|(prop_name, _)| {
+                    if form_data_names.contains(prop_name) {
+                        None
+                    } else {
+                        Some(prop_name)
+                    }
+                })
+                .map(|prop_name| {
+                    // Serialialize text fields.
+                    // Unquoted strings are what servers expect for
+                    // single values, including datetimes.
+                    // We assume they want a json string for nested types.
+                    // The field is not set if it skipped in Serialize.
+                    quote! {
+                        if let Some(v) = body.get(#prop_name)
+                        {
+                            let v = to_form_string(v)?;
+                            #form_ident = #form_ident.text(#prop_name, v);
+                        };
+                    }
+                });
+            form_parts_optional.extend(body_parts);
+        }
+
+        // only build if any param or at least a body without additional parts
+        if form_parts.len() > 0
```

Comment:
`!form_parts.is_empty()` here, and on next line, and `let mutt` a few lines lower.